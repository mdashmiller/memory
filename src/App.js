import React, { Component } from 'react'

class App extends Component {

	state = {
		clicked: false
	}

	// component methods

	changeColor = () =>
		this.setState({ clicked: true })

	render() {
		return (
			<div
				className={this.state.clicked ? 'animate' : undefined}
				onClick={() => this.changeColor()}
			>
			</div>
		)
	}
}

export default App
