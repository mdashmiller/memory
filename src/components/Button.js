import styled from 'styled-components'

const Button = styled.button`
	height: 40px;
	width: 80px;
	text-align: center;
	border-radius: 7px;
	margin: 0 auto;
	display: block;
	border: none;
	font-size: 1em;
	font-family: Arial;
	background: yellow;
	color: red;
	font-weight: 700;
`

export default Button
