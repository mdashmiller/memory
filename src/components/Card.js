import styled from 'styled-components'

const Card = styled.div`
	height: 80px;
	width: 72px;
	border-radius: 5px;
	background-color: red;
`

export default Card
