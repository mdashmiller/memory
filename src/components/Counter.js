import React from 'react'
import styled from 'styled-components'

const CounterLabel = styled.h4`
	color: #bbc4ef;
	margin: 8px 0 0;
`

const StyledCounter = styled.div`
	width: 54px;
    height: 36px;
    margin: 4px auto 0;
    font-size: 32px;
    color: #bbc4ef;
    border: 2px solid #bbc4ef;
    border-radius: 4px;
`

const Counter = ({ moves }) => 
	<div>
		<CounterLabel id="label">MOVES</CounterLabel>
		<StyledCounter id="counter">
			{moves}
		</StyledCounter>
	</div>

export default Counter
