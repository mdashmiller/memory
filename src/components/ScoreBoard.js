import React from 'react'
import styled from 'styled-components'
import emojisArr from '../assets/images-arr'
import PropTypes from 'prop-types'

const EmptyCircle = styled.div`
	width: 22px;
	height: 22px;
	margin: 10px 0 0;
	display: inline-block;
	border: 1px solid #bbc4ef;
	border-radius: 13px;
`

const LilSmile = styled.img`
	opacity: 1;
	width: 24px;
	height: 24px;
	display: inline-block;
	margin: 10px 0 0;
`

const ScoreBoard = ({ posMatches }) =>
	// empty circles fill-in with smiley faces
	// for each match the user makes
	<div style={{ textAlign: 'center' }}>
		{posMatches.length > 1
			? <LilSmile src={emojisArr[39]} alt="smiley face" className="smile"></LilSmile>
			: <EmptyCircle className="empty"></EmptyCircle>
		}
		{posMatches.length > 3
			? <LilSmile src={emojisArr[39]} alt="smiley face" className="smile"></LilSmile>
			: <EmptyCircle className="empty"></EmptyCircle>
		}
		{posMatches.length > 5
			? <LilSmile src={emojisArr[39]} alt="smiley face" className="smile"></LilSmile>
			: <EmptyCircle className="empty"></EmptyCircle>
		}
		{posMatches.length > 7
			? <LilSmile src={emojisArr[39]} alt="smiley face" className="smile"></LilSmile>
			: <EmptyCircle className="empty"></EmptyCircle>
		}
		{posMatches.length > 9
			? <LilSmile src={emojisArr[39]} alt="smiley face" className="smile"></LilSmile>
			: <EmptyCircle className="empty"></EmptyCircle>
		}
		{posMatches.length > 11
			? <LilSmile src={emojisArr[39]} alt="smiley face" className="smile"></LilSmile>
			: <EmptyCircle className="empty"></EmptyCircle>
		}
	</div>

ScoreBoard.propTypes = {
	posMatches: PropTypes.arrayOf(PropTypes.number).isRequired
}

export default ScoreBoard
