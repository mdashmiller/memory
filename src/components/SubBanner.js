import React from 'react'
import styled from 'styled-components'

const StyledSubBanner = styled.div`
	grid-column: 2 / span 2;
	/* border: 1px solid black; */
	height: 24px;
	text-align: center;
	font-size: 19px;
	font-weight: 700;
	color: #fccdd3;
`

const displayScore = score =>
	score ? score : '?'

const SubBanner = ({ bestScore }) =>
	<StyledSubBanner id="sub-banner">
		Best Score: {displayScore(bestScore)}
	</StyledSubBanner>

export default SubBanner
