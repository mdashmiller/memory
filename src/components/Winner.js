import React from 'react'
import styled from 'styled-components'
import Button from './Button'

const Banner = styled.div`
	opacity: 0.9;
    background: black;
    height: 290px;
    width: 320px;
    text-align: center;
    color: #fccdd3;
    margin: 14px auto;
    position: absolute;
    border-radius: 4px
`

const BannerTitle = styled.p`
	font-size: 44px;
    font-weight: 700;
    margin-top: 20px;
    margin-bottom: 63px;
`

const Score = styled.p`
	font-size: 26px;
    margin: 0 auto;
    color: #62bcfa;
`

const BannerSubTitle = styled.p`
	font-size: 30px;
	margin: 0 auto;
`

const Winner = ({
	moves,
	newBest,
	replay
}) =>
	<Banner>
		<BannerTitle>Winner!!!</BannerTitle>
		<Score>Moves: {moves}</Score>
		{newBest &&
			<BannerSubTitle className="flicker">
				New Record!
			</BannerSubTitle>
		}
		<Button
			win
			onClick={replay}
		>
			Play Again?
		</Button>
	</Banner>

export default Winner
