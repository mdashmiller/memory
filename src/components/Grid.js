import React, { Component } from 'react'
import styled from 'styled-components'
import Card from './Card'
import Button from './Button'
import emojisArr from '../assets/images-arr'

// styled component
const StyledGrid = styled.div`
	margin: 20px auto; 
	width: 300px;
	display: grid;
	grid-template-columns: repeat(4, 1fr);
	grid-gap: 4px;
`

class Grid extends Component {

	state = {
		card1Clicked: false,
		card2Clicked: false,
		card3Clicked: false,
		card4Clicked: false,
		card5Clicked: false,
		card6Clicked: false,
		card7Clicked: false,
		card8Clicked: false,
		card9Clicked: false,
		card10Clicked: false,
		card11Clicked: false,
		card12Clicked: false,
		img1Clicked: false,
		img2Clicked: false,
		img3Clicked: false,
		img4Clicked: false,
		img5Clicked: false,
		img6Clicked: false,
		img7Clicked: false,
		img8Clicked: false,
		img9Clicked: false,
		img10Clicked: false,
		img11Clicked: false,
		img12Clicked: false,
		randomInts: []
	}

	// component methods

	handleClick = number => {
		// routes interaction with 'reset'
		// button to proper channels
		this.changeColor(number)
		this.fadeIn(number)
	}

	changeColor = number => {
		// sets the state of each card that is clicked
		const key = `card${number}Clicked`
		this.setState({ [key]: true })
	}

	fadeIn = number => {
		// sets the state for the img on each card clicked
		const key = `img${number}Clicked`
		this.setState({ [key]: true })
	}

	setRandomInts = () => {
		// creates an array of random integer duplicate-pairs
		// equal in length to the number of cards and
		// sets state accordingly
		let { randomInts } = this.state
		randomInts = []//[8,8]
		for (let x = 0; x < 6; x++) {
				//const randomInt = this.getRandomInt(emojisArr.length)
				//let randomInt = this.getRandomInt(emojisArr.length)
				//this.checkArrayForInt(randomInt, randomInts)
				const randomInt = this.createUniqueInt(randomInts)//
				randomInts.push(randomInt)
				randomInts.push(randomInt)
		}
		//this.randomizeArray(randomInts)
		this.setState({ randomInts })
	}

	getRandomInt = max =>
		// creates an almost-random integer
		// equal to or less than the highest index
		// of emojisArr
		Math.floor(Math.random() * Math.floor(max))

	createUniqueInt = arr => {//[8,8]
		const randomInt = this.getRandomInt(emojisArr.length)//8
		if (arr.includes(randomInt)) {
			this.createUniqueInt(arr)
		} else {
			return randomInt
		}
	}

	randomizeArray = arr => {
		const randomizedArr = arr.sort(
			() => (0.5 - Math.random())
		)
		this.setState({ randomInts: randomizedArr })
	}

	shuffle = () => {
		// resets state to its intial configuration and
		// calls for the generation of a new
		// array of random ints
		this.setState({
			card1Clicked: false,
			card2Clicked: false,
			card3Clicked: false,
			card4Clicked: false,
			card5Clicked: false,
			card6Clicked: false,
			card7Clicked: false,
			card8Clicked: false,
			card9Clicked: false,
			card10Clicked: false,
			card11Clicked: false,
			card12Clicked: false,
			img1Clicked: false,
			img2Clicked: false,
			img3Clicked: false,
			img4Clicked: false,
			img5Clicked: false,
			img6Clicked: false,
			img7Clicked: false,
			img8Clicked: false,
			img9Clicked: false,
			img10Clicked: false,
			img11Clicked: false,
			img12Clicked: false,
			randomInts: []
		})
		this.setRandomInts()
	}
		
	// lifecycle methods

	componentDidMount() {
		this.setRandomInts()
	}
		
	render() {
		const { randomInts } = this.state
		console.log(randomInts)
		console.log(typeof randomInts[0])
		return (
			<StyledGrid>
				<Card
					className={this.state.card1Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(1)}
				>
					<img
						src={emojisArr[randomInts[0]]} 
						alt="random emoji"
						className={this.state.img1Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card2Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(2)}
				>
					<img
						src={emojisArr[randomInts[1]]} 
						alt="random emoji"
						className={this.state.img2Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card3Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(3)}
				>
					<img
						src={emojisArr[randomInts[2]]} 
						alt="random emoji"
						className={this.state.img3Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card4Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(4)}
				>
					<img
						src={emojisArr[randomInts[3]]} 
						alt="random emoji"
						className={this.state.img4Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card5Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(5)}
				>
					<img
						src={emojisArr[randomInts[4]]} 
						alt="random emoji"
						className={this.state.img5Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card6Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(6)}
				>
					<img
						src={emojisArr[randomInts[5]]} 
						alt="random emoji"
						className={this.state.img6Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card7Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(7)}
				>
					<img
						src={emojisArr[randomInts[0]]} 
						alt="random emoji"
						className={this.state.img7Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card8Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(8)}
				>
					<img
						src={emojisArr[randomInts[0]]} 
						alt="random emoji"
						className={this.state.img8Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card9Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(9)}
				>	
					<img
						src={emojisArr[randomInts[0]]} 
						alt="random emoji"
						className={this.state.img9Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card10Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(10)}
				>	
					<img
						src={emojisArr[randomInts[0]]} 
						alt="random emoji"
						className={this.state.img10Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card11Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(11)}
				>	
					<img
						src={emojisArr[randomInts[0]]} 
						alt="random emoji"
						className={this.state.img11Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<Card
					className={this.state.card12Clicked ? 'color-change' : undefined}
					onClick={() => this.handleClick(12)}
				>	
					<img
						src={emojisArr[randomInts[0]]} 
						alt="random emoji"
						className={this.state.img12Clicked ? 'fade-in' : undefined}
					/>
				</Card>
				<div className="button-row placeholder"></div>
				<Button
					className="button-row"
					onClick={() => this.shuffle()}
				>
					Reset
				</Button>
				<div className="button-row placeholder"></div>
				<div className="button-row placeholder"></div>
			</StyledGrid>
		)
	}
}

export default Grid
