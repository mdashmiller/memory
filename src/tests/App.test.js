import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import chai, { expect } from 'chai'
import { spy } from 'sinon'

import App from '../App'
import Grid from '../components/Grid'
import Card from '../components/Card'
import Counter from '../components/Counter'
import ScoreBoard from '../components/ScoreBoard'
import Button from '../components/Button'
import Winner from '../components/Winner'
import SubBanner from '../components/SubBanner'

Enzyme.configure({ adapter: new Adapter() })

global.jestExpect = global.expect
global.expect = chai.expect

describe('App', () => {

	it('renders without crashing', () => {
	  const div = document.createElement('div')
	  ReactDOM.render(<App />, div)
	  ReactDOM.unmountComponentAtNode(div)
	})

	// test('has a valid snapshot', () => {
	// 	const component = renderer.create(
	// 		<App />
	// 	)
	// 	let tree = component.toJSON()
	// 	jestExpect(tree).toMatchSnapshot()
	// })

	// it('renders the <Winner /> component when all the matches are found', () => {
	// 	const posMatches = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
	// 	const wrapper = shallow(<App posMatches={posMatches} />)
	// 	expect(wrapper.find('Winner')).to.have.lengthOf(1)
	// })

	it('renders 1 <Grid /> component', () => {
		const wrapper = shallow(<App />)
		expect(wrapper.find(Grid)).to.have.lengthOf(1)
	})
	
	it('renders 3 div.button-row', () => {
		const wrapper = shallow(<App />)
		expect(wrapper.find('div.button-row')).to.have.lengthOf(3)
	})

	it('renders 1 div.counter', () => {
		const wrapper = shallow(<App />)
		expect(wrapper.find('div.counter')).to.have.lengthOf(1)
	})

	it('renders 1 <Counter /> component', () => {
		const wrapper = shallow(<App />)
		expect(wrapper.find(Counter)).to.have.lengthOf(1)
	})

	it('renders 1 div.score', () => {
		const wrapper = shallow(<App />)
		expect(wrapper.find('div.score')).to.have.lengthOf(1)
	})

	it('renders 1 <ScoreBoard /> component', () => {
		const wrapper = shallow(<App />)
		expect(wrapper.find(ScoreBoard)).to.have.lengthOf(1)
	})

	it('renders 1 <Button /> component', () => {
		const wrapper = shallow(<App />)
		expect(wrapper.find(Button)).to.have.lengthOf(1)
	})

	it('renders 1 <SubBanner /> component', () => {
		const wrapper = shallow(<App />)
		expect(wrapper.find(SubBanner)).to.have.lengthOf(1)
	})

})
