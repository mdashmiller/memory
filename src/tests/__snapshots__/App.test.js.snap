// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div>
  <h1>
    Do You Remember?
  </h1>
  <div
    className="main"
  >
    <div
      className="sc-gZMcBi dChKYF"
    >
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="crying.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="rolling-eyes.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="very-sad.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="hungry.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="sweat-smile.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="rolling-eyes.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="crying.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="hungry.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="flushed.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="flushed.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="sweat-smile.png"
        />
      </div>
      <div
        className="sc-bdVaJa kPlaWI"
        onClick={[Function]}
      >
        <img
          alt="random emoji"
          src="very-sad.png"
        />
      </div>
      <div
        className="button-row counter"
      >
        <div>
          <h4
            className="sc-bwzfXH iOSHLe"
          >
            MOVES
          </h4>
          <div
            className="sc-htpNat dVIvjG"
          >
            00
          </div>
        </div>
      </div>
      <div
        className="button-row score"
      >
        <div
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          <div
            className="sc-bxivhb LfNcR"
          />
          <div
            className="sc-bxivhb LfNcR"
          />
          <div
            className="sc-bxivhb LfNcR"
          />
          <div
            className="sc-bxivhb LfNcR"
          />
          <div
            className="sc-bxivhb LfNcR"
          />
          <div
            className="sc-bxivhb LfNcR"
          />
        </div>
      </div>
      <div
        className="button-row"
      >
        <button
          className="sc-EHOje jRqDu"
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
      <div
        className="sc-iwsKbI bujAyb"
      >
        Best Score: 
        ?
      </div>
    </div>
  </div>
</div>
`;
